lib/QueryValidationVisitor.js

```diff
@@ -0,0 +1,229 @@
+        )
+    }
+    // no parent set as we are at the top of tree
+    this.currentTypeInfo = { typeDef }
+  }
+
+  onFragmentEnter (node) {
+    const newTypeDef = typeFromAST(this.context.getSchema(), node.typeCondition)
+    this.currentTypeInfo = { parent: this.currentTypeInfo, typeDef: newTypeDef }
+  }
+
+  onFragmentLeave (node) {
+    this.currentTypeInfo = this.currentTypeInfo.parent
+  }
+
+  onFieldEnter (node) {
+    // logDebug('onFieldEnter ', JSON.stringify(node))
+    // logDebug('onFieldEnter this.currentTypeInfo.typeDef ', JSON.stringify(this.currentTypeInfo.typeDef))
+    // prepere current field info, so onArgumentEnter() can use it
+    this.currentField = node
+    this.currentrFieldDef = this.currentTypeInfo.typeDef.getFields()[node.name.value]
+    if (this.currentrFieldDef) {
+      const newTypeDef = getNamedType(this.currentrFieldDef.type)
+      this.currentTypeInfo = { parent: this.currentTypeInfo, typeDef: newTypeDef }
+    } else {
+      // logDebug('onSelectionSetEnter is part of the Introspection query')
+      return BREAK
+    }
+  }
+
+  onFieldLeave (node) {
+    this.currentTypeInfo = this.currentTypeInfo.parent
+  }
+
+  onArgumentEnter (arg) {
+    // logDebug('onArgumentEnter ', JSON.stringify(arg))
+    const argName = arg.name.value
+    // look for directive and validate argument if directive found
+    const argTypeDef = this.currentrFieldDef.args.find(d => d.name === argName)
+    const value = valueFromAST(arg.value, argTypeDef.type, this.variableValues)
+
+    // nothing to validate
+    if (!value) return
+
+    let variableName
+    if (arg.value.kind === Kind.VARIABLE) variableName = arg.value.name.value
+
+    let valueTypeDef = argTypeDef.type
+    if (isNonNullType(valueTypeDef)) valueTypeDef = valueTypeDef.ofType
+
+    if (isInputObjectType(valueTypeDef)) {
+      const inputObjectTypeDef = getNamedType(valueTypeDef)
+      validateInputTypeValue(this.context, inputObjectTypeDef, argName, variableName, value, this.currentField, variableName)
+    } else if (isListType(valueTypeDef)) {
+      validateArrayTypeValue(this.context, valueTypeDef, argTypeDef, value, this.currentField, argName, variableName, variableName)
+    } else {
+      const fieldNameForError = variableName || this.currentField.name.value + '.' + argName
+      validateScalarTypeValue(this.context, this.currentField, argTypeDef, valueTypeDef, value, variableName, argName, fieldNameForError, '')
+    }
+  }
+}
+
+function validateScalarTypeValue (context, currentQueryField, typeDefWithDirective, valueTypeDef, value, variableName, argName, fieldNameForError, errMessageAt) {
+  const directiveArgumentMap = getDirectiveValues(constraintDirectiveTypeDefsObj, typeDefWithDirective.astNode)
+  if (directiveArgumentMap) {
+    const st = getScalarType(valueTypeDef).scalarType
+    const valueDelim = st === GraphQLString ? '"' : ''
+    try {
```

Comment:
I generally prefer spacing after variable decelerations, makes it easier to read e.g.

```
const st = getScalarType(valueTypeDef).scalarType
const valueDelim = st === GraphQLString ? '"' : ''

try {
```

Same with if statements as well; this function seems quite compact

```
      let error

      if (variableName) {
        error = new ValidationError(fieldNameForError, `Variable "$${variableName}" got invalid value ${valueDelim}${value}${valueDelim}${errMessageAt}. ` + e.message, e.context)
      } else {
        error = new ValidationError(fieldNameForError, `Argument "${argName}" of "${currentQueryField.name.value}" got invalid value ${valueDelim}${value}${valueDelim}${errMessageAt}. ` + e.message, e.context)
      }

      error.originalError = e
```

Reply:
Sure, I'm pretty new in JavaScript world so do not know best practices applied here :-D